### ISSUE TITLE
`StructuralPartialEq` pat check does not check hr regions

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust#160122]
The following code
```rust
#![feature(structural_match)]

struct Thing;

trait IsStatic {}
impl<'a: 'static> IsStatic for &'a () {}

impl std::marker::StructuralPartialEq for Thing where for<'a> &'a (): IsStatic {}

impl PartialEq for Thing
where
    for<'a> &'a (): IsStatic,
{
    fn eq(&self, _: &Thing) -> bool {
        panic!()
    }
}

const A: Thing = Thing;

fn main() {
    let A = Thing;
}
```
built with `--edition=2024`
compiled with the old solver:
```
<no output>
```
but failed to compile with the new solver:
```
error: constant of non-structural type `Thing` in a pattern
  --> ./main_corpus/88596606350047689476980845282776363502.rs:22:9
   |
 3 | struct Thing;
   | ------------ `Thing` must be annotated with `#[derive(PartialEq)]` to be usable in patterns
...
19 | const A: Thing = Thing;
   | -------------- constant defined here
...
22 |     let A = Thing;
   |         ^ constant of non-structural type

error: aborting due to 1 previous error


```

version inforamtion
```
rustc 1.99.0-nightly (4fefe3656 2026-07-27)
binary: rustc
commit-hash: 4fefe3656700e5528edddc2336667818917dfc70
commit-date: 2026-07-27
host: x86_64-unknown-linux-gnu
release: 1.99.0-nightly
LLVM version: 22.1.8

```

**Comment by lcnr** [comment]
intended breakage. the old solver `StructuralPartialEq` does not detect higher-ranked region errors here 😅
